File: JavaLibrary/src/test/java/com/flat/simplex/support/ContextSupport.java
package com.flat.simplex.support;

import com.flat.simplex.lexer.Token;
import com.flat.simplex.parser.logic.Context;
import com.flat.simplex.parser.logic.error.Error;
import org.junit.jupiter.api.Assertions;
import org.opentest4j.AssertionFailedError;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Objects;

public class ContextSupport {

    public static void assertFields(Context context, String... names) {
        List<String> contextFields = context.getAllFieldNames().stream().sorted().toList();
        List<String> fieldNames = Arrays.stream(names).sorted().toList();

        boolean dif = contextFields.size() != fieldNames.size();
        if (!dif) {
            for (int i = 0; i < contextFields.size(); i++) {
                if (!Objects.equals(fieldNames.get(i), contextFields.get(i))) {
                    dif = true;
                    break;
                }
            }
        }
        if (dif) {
            String expected = "";
            for (String desc : fieldNames) {
                expected += desc + "\n";
            }

            String actual = "";
            for (String desc : contextFields) {
                actual += desc + "\n";
            }
            throw new AssertionFailedError("Invalid Syntax Errors", expected, actual);
        }
    }

    public static void assertErrors(Context context, String... descriptions) {
        ArrayList<Error> errors = context.getErrors();

        boolean dif = errors.size() != descriptions.length;
        if (!dif) {
            for (int i = 0; i < errors.size(); i++) {
                if (!Objects.equals(descriptions[i], errors.get(i).getDescription())) {
                    dif = true;
                    break;
                }
            }
        }
        if (dif) {
            String expected = "";
            for (String desc : descriptions) {
                expected += desc + "\n";
            }

            String actual = "";
            for (Error desc : errors) {
                actual += desc.getDescription() + "\n";
            }
            throw new AssertionFailedError("Invalid Syntax Errors", expected, actual);
        }

    }
}

File: SimplexApp/simplex/support/ContextSupport.js
var AssertError;

class ContextSupport {
    static assertFields (context, ...names) {
        let contextFields = context.getAllFieldNames().sort();
        let fieldNames = names.sort();
        let dif = contextFields.length !== fieldNames.length;
        if (!dif) {
            for (let i = 0; i < contextFields.length; i++) {
                if (fieldNames[i] !== contextFields[i]) {
                    dif = true;
                    break;
                }
            }
        }
        if (dif) {
            let expected = '';
            for (const desc of fieldNames) {
                expected += desc + '\n';
            }
            let actual = '';
            for (const desc of contextFields) {
                actual += desc + '\n';
            }
            throw new AssertError('Invalid Syntax Errors', expected, actual);
        }
    }

    static assertErrors(context, ...descriptions) {
        let errors = context.getErrors();
        let dif = errors.length !== descriptions.length;
        if (!dif) {
            for (let i = 0; i < errors.length; i++) {
                if (descriptions[i] !== errors[i].getDescription()) {
                    dif = true;
                    break;
                }
            }
        }
        if (dif) {
            let expected = '';
            for (const desc of descriptions) {
                expected += desc + '\n';
            }
            let actual = '';
            for (const desc of errors) {
                actual += desc.getDescription() + '\n';
            }
            throw new AssertError('Invalid Syntax Errors', expected, actual);
        }
    }
}

module.exports = ContextSupport;
AssertError = require("simplex/support/AssertError.js");